Change getTgtMemIntrinsic interface with returning a pair

HankChang736 opened the pull request:
* This patch change the interface of getTgtMemIntrinsic from return "bool" into return "std::pair<std::optional<MemIntrinsicInfo>,std::optional<SmallVectorImpl<InterestingMemoryOperand>>>" to allow Asan or EarlyCSE to choose the data structure they need with the same TTI hook.

HankChang736 committed:
Change getTgtMemIntrinsic interface with returning a pair

* This patch change the interface of getTgtMemIntrinsic from return "bool" into return
"std::pair<std::optional<MemIntrinsicInfo>,std::optional<SmallVectorImpl<InterestingMemoryOperand>>>"
to allow Asan or EarlyCSE to choose the data structure they need with the same TTI hook.

kito-cheng commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -500,29 +500,32 @@ unsigned GCNTTIImpl::getMaxInterleaveFactor(ElementCount VF) const {
     auto *Volatile = dyn_cast<ConstantInt>(Inst->getArgOperand(4));
     if (!Ordering || !Volatile)
-      return false; // Invalid.
+      return std::make_pair(std::nullopt, std::nullopt);
```
```suggestion
      return std::make_pair(std::nullopt, std::nullopt); // Invalid.
```

kito-cheng commented on llvm/lib/Transforms/Scalar/EarlyCSE.cpp:
```diff
@@ -834,6 +832,11 @@ class EarlyCSE {
+                  std::optional<SmallVector<InterestingMemoryOperand, 1>>>
+            MemInfo = TTI.getTgtMemIntrinsic(II);
+        if (MemInfo.first != std::nullopt)
+          Info = *MemInfo.first;
```
This part...

kito-cheng commented on llvm/lib/Transforms/Scalar/EarlyCSE.cpp:
```diff
@@ -808,8 +808,6 @@ class EarlyCSE {
         IntrID = II->getIntrinsicID();
-        if (TTI.getTgtMemIntrinsic(II, Info))
-          return;
         if (isHandledNonTargetIntrinsic(IntrID)) {
```
Should stay within this if-block since it check `IntrinsicInst *II = dyn_cast<IntrinsicInst>(Inst)`

HankChang736 replied:
Do you mean stay within the if-block `if (IntrinsicInst *II = dyn_cast<IntrinsicInst>(Inst))`?
If so, it stays within this if-block that checks `IntrinsicInst *II = dyn_cast<IntrinsicInst>(Inst)`.

kito-cheng replied:
Oh, my bad, I misreading the indention, but I would still suggest move those line here to provide better readability for the diff.

HankChang736 replied:
Ok, that sure sounds better for the readability.
I just moved it above.

HankChang736 committed:
Address Kito's comment

HankChang736 committed:
Address Kito's 2nd comment

topperc commented on llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:
```diff
@@ -975,9 +975,10 @@ class TargetTransformInfoImplBase {
-                                  MemIntrinsicInfo &Info) const {
-    return false;
+  virtual std::pair<std::optional<MemIntrinsicInfo>,
+                    std::optional<SmallVector<InterestingMemoryOperand, 1>>>
```
Do we need an optional around the SmallVector? Can we use an empty vector to represent no interesting operand?

Looks like EarlyCSE will ignore MemIntrinsicInfo with a null `PtrVal`. Can we use that instead of an optional?

HankChang736 replied:
I think use an empty vector to represent no interesting operand is a great idea, in this way we don't need to use optional for SmallVector.

Use a null `PtrVal` to check MemIntrinsicInfo might be fine, I just checked the target that uses getTgtMemIntrinsic and find that all of them fill `PtrVal` and EarlyCSE/LoopStrengthReduce also ignore `PtrVal` with nullptr.

I will try to modify it with dropping optional.

HankChang736 replied:
```suggestion
  std::pair<MemIntrinsicInfo,
            SmallVector<InterestingMemoryOperand, 1>>
```
The interface after dropping `std::optional` will look like this.

topperc commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -32,6 +32,7 @@
 #include "llvm/Support/BranchProbability.h"
 #include "llvm/Support/Compiler.h"
 #include "llvm/Support/InstructionCost.h"
+#include "llvm/Transforms/Instrumentation/AddressSanitizerCommon.h"
```
I don't think Analysis should include a header from Transforms. The Transforms library has a dependency on the Analysis library.

HankChang736 replied:
Oh I see. The reason I include this header from Transforms is to use `InterestingMemoryOperand`, maybe move `InterestingMemoryOperand` from Transforms header to `Analysis/InterestingMemoryOperand.h` and include it in `Transforms/Instrumentation/AddressSanitizerCommon.h` ?

topperc replied:
That sounds good to me.

HankChang736 committed:
Move InterestingMemoryOperand from Transforms header to Analysis

HankChang736 commented on llvm/lib/Transforms/Scalar/EarlyCSE.cpp:
```diff
@@ -808,8 +808,13 @@ class EarlyCSE {
+        std::pair<std::optional<MemIntrinsicInfo>,
+                  std::optional<SmallVector<InterestingMemoryOperand, 1>>>
+            MemInfo = TTI.getTgtMemIntrinsic(II);
+        if (MemInfo.first != std::nullopt) {
```
```suggestion
        if (MemInfo.first.PtrVal) {
```
Check `PtrVal` is null or not to see whether the target fill MemIntrinsicInfo.

HankChang736 commented on llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:
```diff
@@ -1018,9 +1018,11 @@ static bool isAddressUse(const TargetTransformInfo &TTI,
+      std::pair<std::optional<MemIntrinsicInfo>,
+                std::optional<SmallVector<InterestingMemoryOperand, 1>>>
+          MemInfo = TTI.getTgtMemIntrinsic(II);
+      if (MemInfo.first != std::nullopt) {
```
```suggestion
      if (MemInfo.first.PtrVal) {
```
Same here

HankChang736 commented on llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:
```diff
@@ -1074,10 +1076,12 @@ static MemAccessTy getAccessType(const TargetTransformInfo &TTI,
+      std::pair<std::optional<MemIntrinsicInfo>,
+                std::optional<SmallVector<InterestingMemoryOperand, 1>>>
+          MemInfo = TTI.getTgtMemIntrinsic(II);
+      if (MemInfo.first != std::nullopt && MemInfo.first->PtrVal) {
```
```suggestion
      if (MemInfo.first.PtrVal) {
```
Check `PtrVal` here too.

HankChang736 commented on llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:
```diff
@@ -1568,6 +1572,15 @@ void AddressSanitizer::getInterestingMemoryOperands(
+                  std::optional<SmallVector<InterestingMemoryOperand, 1>>>
+            MemInfo;
+        MemInfo = TTI->getTgtMemIntrinsic(II);
+        if (MemInfo.second != std::nullopt)
```
```suggestion
        if (!MemInfo.second.empty())
```
Check the SmallVector is empty or not.

HankChang736 committed:
Drop std::optional in getTgtMemIntrinsic return type

This commit drops std::optional in getTgtMemIntrinsic return type,
use nullptr in PtrVal to identify empty MemIntrinsicInfo and empty()
to identify empty SmallVector<InterestingMemoryOperand, 1>.

HankChang736 commented:
@topperc 
I just update the return type in getTgtMemIntrinsic() with dropping std::optional, now it use `nullptr` in `PtrVal` to check empty MemIntrinsicInfo and use empty() to check empty SmallVector.

The commit that include this update is [f2abced960e7be54d8eb3aa6557dbc472c4f40f2](https://github.com/HankChang736/llvm-project/pull/1/commits/f2abced960e7be54d8eb3aa6557dbc472c4f40f2)

topperc commented on llvm/include/llvm/Analysis/InterestingMemoryOperand.h:
```diff
@@ -0,0 +1,55 @@
+//===--------- Definition of the InterestingMemoryOperand class -*- C++
```
Don't wrap this line

topperc commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1696,8 +1697,8 @@ class TargetTransformInfo {
   /// if false is returned.
-  LLVM_ABI bool getTgtMemIntrinsic(IntrinsicInst *Inst,
-                                   MemIntrinsicInfo &Info) const;
+  LLVM_ABI std::pair<MemIntrinsicInfo, SmallVector<InterestingMemoryOperand, 1>>
```
Could we make InterestingMemoryOperand a member of MemIntrinsicInfo and keep the original signature?

HankChang736 replied:
I’m worried the original signature by returning `bool` might be ambiguous for targets that never produce `SmallVector<InterestingMemoryOperand, 1>`.
With `SmallVector<InterestingMemoryOperand, 1>` embedded in `MemIntrinsicInfo`, an empty vector can mean either:

1. The target didn’t implement getTgtMemIntrinsic.

2. The target did implement it and there are truly no `SmallVector<InterestingMemoryOperand, 1>`.

topperc replied:
Aren't those both possibilities if you return a SmallVector as part of a pair? Is the difference between those two cases meaningful for the current use case? Should we go back to the std::optional?

HankChang736 replied:
Yes, it is. 

But returning a pair essentially makes the Transform passes that use `getTgtMemIntrinsic` directly check whether the returned `SmallVector` is empty or not with an `if` condition. On the other hand, embedding `SmallVector` inside `MemIntrinsicInfo` while keeping the original signature `bool getTgtMemIntrinsic(IntrinsicInst *Inst, MemIntrinsicInfo &Info)` introduces another subtlety: since returning true only guarantees that `MemIntrinsicInfo` is filled, but some targets like AArch64, PowerPC and AMDGPU don't fill `SmallVector`. 
This means that passes like `AddressSanitizer`, which only rely on `SmallVector`, could end up accessing it even when it is empty, leading to ambiguity or potential misuse. As a result, we would still need to double-check whether the `SmallVector` is empty to ensure the target actually provided the corresponding data.

I think making `InterestingMemoryOperand` a member of `MemIntrinsicInfo` is also an approach. Since it keeps the original implementation, but it requires a double-check like this in AddressSanitizer pass :

```cpp
if (TTI.getTgtMemIntrinsic(Inst, Info) && !Info.Interesting.empty())
```
Also it needs to check the `PtrVal` in `EarlyCSE` to ensure whether the target really fills the information for it since those targets that support `getTgtMemIntrinsic` all fills `PtrVal`.

```cpp
if (TTI.getTgtMemIntrinsic(Inst, Info) && Info.PtrVal)
```

HankChang736 replied:
I think making InterestingMemoryOperand a member of MemIntrinsicInfo is also an approach. Since it keeps the original implementation, but it requires a double-check like this in AddressSanitizer pass :

```cpp
if (TTI.getTgtMemIntrinsic(Inst, Info) && !Info.Interesting.empty())
```
Also it needs to check the `PtrVal` in `EarlyCSE` to ensure whether the target really fills the information for it since those targets that support `getTgtMemIntrinsic` all fills `PtrVal`.

```cpp
if (TTI.getTgtMemIntrinsic(Inst, Info) && Info.PtrVal)
```

topperc replied:
Doesn't EarlyCSE check `ParseMemoryInst::isValid()` to check if the PtrVal is null before doing anything with it.

> This means that passes like AddressSanitizer, which only rely on SmallVector, could end up accessing it even when it is empty, leading to ambiguity or potential misuse. As a result, we would still need to double-check whether the SmallVector is empty to ensure the target actually provided the corresponding data.

Doesn't AddressSanitizer need to check the vector is non-empty before using it regardless of where it is stored?

HankChang736 replied:
Ah, I see — that makes sense now.

For EarlyCSE, `ParseMemoryInst::isValid()` already ensures that PtrVal is non-null before using it, so we don’t actually need an extra check there.
And for AddressSanitizer, it explicitly verifies that the vector is non-empty before using it. So it looks like we can indeed make InterestingMemoryOperand a member of MemIntrinsicInfo safely.

Thanks for pointing this out — it clears up my earlier concern.

topperc commented on llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:
```diff
@@ -2983,7 +2994,8 @@ bool AddressSanitizer::suppressInstrumentationSiteForDebug(int &Instrumented) {
 bool AddressSanitizer::instrumentFunction(Function &F,
-                                          const TargetLibraryInfo *TLI) {
+                                          const TargetLibraryInfo *TLI,
+					  const TargetTransformInfo *TTI) {
```
weird formatting

topperc commented on llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:
```diff
@@ -1018,9 +1018,10 @@ static bool isAddressUse(const TargetTransformInfo &TTI,
-        if (IntrInfo.PtrVal == OperandVal)
+      std::pair<MemIntrinsicInfo, SmallVector<InterestingMemoryOperand, 1>>
+          MemInfo = TTI.getTgtMemIntrinsic(II);
+      if (MemInfo.first.PtrVal != nullptr) {
```
Is OperandVal always non-null? If so, it could never match a null `MemInfo.first.PtrVal`.

If we do this check, then use `&&` to join the two ifs instead of nesting them.

HankChang736 replied:
I'm not sure if OperandVal is always non-null here. Maybe we keep this check since we make SmallVector<InterestingMemoryOperand> a member of MemIntrinsicInfo and keep the original signatures.

HankChang736 committed:
[TTI] Make SmallVector<InterestingMemoryOperand> a member of MemIntrinsicInfo

This commit make SmallVector<InterestingMemoryOperand> a member of MemIntrinsicInfo so that
we can keep the original interface of getTgtMemIntrinsic.

HankChang736 commented:
@topperc 
I make SmallVector<InterestingMemoryOperand, 1> a member of MemIntrinsic Info, and fix the passes that use getTgtMemIntrinsic in commit [e89d6ba](https://github.com/HankChang736/llvm-project/pull/1/commits/e89d6ba9f1f61cc2d568d8d563d5cc70deba3d63).

topperc commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -501,17 +501,18 @@ unsigned GCNTTIImpl::getMaxInterleaveFactor(ElementCount VF) const {
 
 bool GCNTTIImpl::getTgtMemIntrinsic(IntrinsicInst *Inst,
-                                       MemIntrinsicInfo &Info) const {
+                                    MemIntrinsicInfo &Info) const {
```
Please revert these formatting changes

topperc commented on llvm/lib/Target/PowerPC/PPCTargetTransformInfo.cpp:
```diff
@@ -1024,7 +1024,6 @@ bool PPCTTIImpl::getTgtMemIntrinsic(IntrinsicInst *Inst,
     break;
   }
-
   return false;
```
Please revert this change

topperc commented on llvm/lib/Target/PowerPC/PPCTargetTransformInfo.h:
```diff
@@ -69,8 +69,10 @@ class PPCTTIImpl final : public BasicTTIImplBase<PPCTTIImpl> {
   bool canSaveCmp(Loop *L, BranchInst **BI, ScalarEvolution *SE, LoopInfo *LI,
                   DominatorTree *DT, AssumptionCache *AC,
                   TargetLibraryInfo *LibInfo) const override;
+
```
revert this change

topperc commented on llvm/lib/Transforms/Scalar/EarlyCSE.cpp:
```diff
@@ -808,8 +808,9 @@ class EarlyCSE {
       if (IntrinsicInst *II = dyn_cast<IntrinsicInst>(Inst)) {
         IntrID = II->getIntrinsicID();
-        if (TTI.getTgtMemIntrinsic(II, Info))
+        if (TTI.getTgtMemIntrinsic(II, Info)) {
```
Drop curly braces

topperc commented on llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp:
```diff
@@ -1076,10 +1076,9 @@ static MemAccessTy getAccessType(const TargetTransformInfo &TTI,
       if (TTI.getTgtMemIntrinsic(II, IntrInfo) && IntrInfo.PtrVal) {
-        AccessTy.AddrSpace
-          = IntrInfo.PtrVal->getType()->getPointerAddressSpace();
+        AccessTy.AddrSpace =
```
revert this formatting change

HankChang736 committed:
Fix format with Craig's comment

HankChang736 committed:
Fix format

HankChang736 commented:
@topperc 
The format is changed in commit [9f26e22](https://github.com/HankChang736/llvm-project/pull/1/commits/9f26e22e0ecb245d8d125f9ef4bea3b6f1c520af), [665ebae](https://github.com/HankChang736/llvm-project/pull/1/commits/665ebae5e0b315577574b5b7a230eedc799699f3) and [766d8c5](https://github.com/HankChang736/llvm-project/pull/1/commits/766d8c590bb4c81b5a9352152e0634a71deec576).

HankChang736 committed:
Fix more format

topperc reviewed (approved):
LGTM

HankChang736 commented:
I just created a pr in llvm upstream.

https://github.com/llvm/llvm-project/pull/157863
